### PR TITLE
Vendor — GCP

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
# Google Cloud

### How to test

1. Create a project, a new service account, and its corresponding JSON key file.
2. Export the credentials with `export GOOGLE_APPLICATION_CREDENTIALS_DATA="$(cat /path/to/service-account.json)"`
3. Build and test the provider manually. 😱 

### Low priority pending tasks

* https://github.com/iterative/terraform-provider-iterative/issues/124
* https://github.com/iterative/cml/issues/162 (waiting for quota increase)

***

_Note: this pull request is a follow-up of the boilerplate on #145._

### CODE REVIEW

**Comment by DavidGOrtega** [comment]
Always fails removing the machine
```sh
iterative_cml_runner.runner-gh-20: Refreshing state... [id=iterative-1xusc3iki91we]
iterative_cml_runner.runner-gh-20: Destroying... [id=iterative-1xusc3iki91we]

Error: Failed disposing the machine: googleapi: Error 400: Invalid value for field 'zone': 'us-west'. Unknown zone., invalid

```

File: iterative/gcp/provider.go [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+
+	instanceZone := getRegion(d.Get("region").(string))
+	instanceName := d.Id()
+
+	instanceDeleteOperation, err := service.Instances.Delete(project, instanceZone, instanceName).Do()
+	if err != nil {
+		return err
+	}
+
+	instanceGetOperationCall := service.ZoneOperations.Get(project, instanceZone, instanceDeleteOperation.Name)
```
Comment (DavidGOrtega) [inline iterative/gcp/provider.go:270]: After the machine is destroyed... where is this being run?

**Comment by DavidGOrtega** [comment]
```sh
terraform apply --auto-approve
iterative_cml_runner.runner-gh-20: Refreshing state... [id=iterative-180l50x0y991]
iterative_cml_runner.runner-gh-20: Destroying... [id=iterative-180l50x0y991]

Error: Failed disposing the machine: googleapi: Error 404: The resource 'projects/cml-gcp-test/zones/us-west1-b/instances/iterative-180l50x0y991' was not found, notFound
```
This error is not convenient and makes you cleanup. your env. Other providers just do not wait the response of the operation

**Comment by DavidGOrtega** [comment]
Spot instances seems not to be working

```
terraform apply --auto-approve
iterative_cml_runner.runner-gh-20: Creating...
iterative_cml_runner.runner-gh-20: Still creating... [10s elapsed]

Error: Failed disposing the machine: googleapi: Error 404: The resource 'projects/cml-gcp-test/zones/us-west1-b/instances/iterative-2hjobqowwx33d' was not found, notFound



Error: Failed creating the machine: googleapi: Error 400: Invalid value for field 'resource.scheduling.preemptible': 'true'. Scheduling must have preemptible be false when OnHostMaintenance isn't TERMINATE., invalid
```

**Comment by 0x2b3bfa0** [comment]
Nice catches on https://github.com/iterative/terraform-provider-iterative/pull/156/commits/c4bd9f317fce104f06b31b0df3f6318d0244b1ff and https://github.com/iterative/terraform-provider-iterative/pull/156/commits/af544a2fe39913221c9ba83098dc5686dc5c224f 🎉

**Comment by 0x2b3bfa0** [comment]
I've solved the remaining lapsus and marked your comments as outdated, as I can't use resolved on comments that don't belong to a review. Feel free to check them again.

File: iterative/gcp/provider.go [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+	instanceZone := getRegion(d.Get("region").(string))
+	instanceName := d.Id()
+
+	instanceDeleteOperationCall := service.Instances.Delete(project, instanceZone, instanceName)
+	_, err = waitForOperation(ctx, d.Timeout(schema.TimeoutDelete), instanceDeleteOperationCall.Do)
+	if err != nil {
+		return err
+	}
+
+	firewallIngressDeleteOperationCall := service.Firewalls.Delete(project, instanceName+"-ingress")
```
Comment (DavidGOrtega) [inline iterative/gcp/provider.go:275]: I do not think this is going to happen if the runner disposes the machine, we should use instance-resources to just call one delete

Comment (DavidGOrtega) [reply iterative/gcp/provider.go:275]: Now this code fails every single time

```sh
Error: Failed disposing the machine: googleapi: Error 404: The resource 'projects/cml-gcp-test/zones/us-west1-b/instances/iterative-18m2033bvd8tx' was not found, notFound
```

**Pull request merged**